### PR TITLE
reader: message timeout callbacks

### PR DESCRIPTION
**Pull request by akellehe**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@ dist
 pynsq.egg-info
 build
 docs/_build
+.idea
```
Comment (akellehe) [inline .gitignore]: Probably want to take this out... it's for PyCharm

File: nsq/async.py [hunk 2]
```diff
@@ -468,6 +468,9 @@ def _on_data(self, data, **kwargs):
             self.in_flight += 1
 
             message = protocol.decode_message(data)
+            if self.msg_timeout:
+                message._set_timeout(self)
```
Comment (akellehe) [inline nsq/async.py]: Accessing the protected method here isn't ideal. Trying to convey to the user that `_set_timeout` shouldn't be called in their implementations.

File: nsq/event.py [hunk 3]
```diff
@@ -28,6 +28,7 @@ class InvalidListenerError(Exception):
 TOUCH = 'touch'
 FINISH = 'finish'
 MESSAGE = 'message'
+MESSAGE_TIMEOUT = 'message_timeout'
```
Comment (akellehe) [inline nsq/event.py]: We could shorten this to `TIMEOUT` and `"timeout"` since it's the only timeout event... Would be less descriptive, though.

File: nsq/message.py [hunk 4]
```diff
@@ -55,9 +56,43 @@ def __init__(self, id, body, timestamp, attempts):
...
+        """
+        self.conn = async_conn
+        self._clear_timeout()
+
+        def callback():
+            async_conn.trigger(event.MESSAGE_TIMEOUT, conn=self, message=self)
+            if self.timeout.is_running():
+                self._clear_timeout()
+
+        self.timeout = tornado.ioloop.PeriodicCallback(
```
Comment (akellehe) [inline nsq/message.py]: Using [`tornado.ioloop.PeriodicCallback`][1] instead of [`tornado.ioloop.IOLoop.add_timeout`][2] for the `ms` resolution. We can use `add_timeout` if we pass the delta as a float, though. 

[1]: http://tornadokevinlee.readthedocs.org/en/latest/ioloop.html#tornado.ioloop.PeriodicCallback
[2]: http://tornadokevinlee.readthedocs.org/en/latest/ioloop.html#tornado.ioloop.IOLoop.add_timeout

File: nsq/reader.py [hunk 5]
```diff
@@ -112,6 +112,10 @@ def process_message(message):
     :param max_in_flight: the maximum number of messages this reader will pipeline for processing.
         this value will be divided evenly amongst the configured/discovered nsqd producers
 
+    :param message_timeout_handler: A callable that runs when a message times out. The parameters are
```
Comment (akellehe) [inline nsq/reader.py]: Should we add some docs for `msg_timeout` here?

File: nsq/async.py (hunk 2 above)
Comment (akellehe) [inline nsq/async.py]: As an alternative here we could create a registry for the timeouts here... then we wouldn't have to incorporate them into the state of the message. Another not-so-nice alternative would be to make the method public.

File: nsq/reader.py [hunk 6]
```diff
@@ -275,6 +283,18 @@ def set_message_handler(self, message_handler):
         assert callable(message_handler), 'message_handler must be callable'
         self.message_handler = message_handler
 
+    def set_message_timeout_handler(self, message_timeout_handler):
+        """
+        Assigns the callback method to be executed for each message timeout that occurs. This method can only be used
+            in conjunction with the client-side msg_timeout setting. The client, otherwise, has no awareness of the
+            message's lifespan.
```
Comment (mreiferson) [inline nsq/reader.py]: `nsqd` delivers the configured `msg_timeout` to the client in the response body of the `IDENTIFY` command, so the client always knows what the `msg_timeout` is.

Comment (akellehe) [inline nsq/reader.py]: Awesome! That makes this feature _much_ more intuitive.

File: nsq/message.py (hunk 4 above)
Comment (mreiferson) [inline nsq/message.py]: I feel like this should be implemented similar to the way the other message-based events are implemented, specifically that it triggers an event here that the parent connection is listening for, that can then call the appropriate callbacks.  I think that would reduce the state we're tracking here?

File: nsq/async.py (hunk 2 above)
Comment (mreiferson) [inline nsq/async.py]: (and would probably obviate the need for this method at all)

**Comment by akellehe** [comment]
@mreiferson what do you think about this approach? 
* Removed `_set_timeout` method from :class:`pynsq.message.Message`
* Added `timed_out` public attribute there.
* Added private `__timeouts` registry to :class:`nsq.async.AsyncConn`
* Changed to `add_timeout` interface and moved to :class:`nsq.async.AsyncConn`

File: nsq/message.py [hunk 7]
```diff
@@ -51,13 +51,22 @@ class Message(event.EventedMixin):
...
+        self.timed_out = False
 
         super(Message, self).__init__()
 
+    def is_alive(self):
+        """
+        Indicates whether or not this :class:`nsq.message.Message` instance is timed out based on the
+            msg_timeout parameter.
+        """
+        return self.timed_out
```
Comment (akellehe) [inline nsq/message.py]: We may want to use `not self.timed_out and not self._has_responded` here.

File: nsq/message.py [hunk 8]
```diff
@@ -51,13 +51,22 @@ class Message(event.EventedMixin):
...
+        self.timed_out = False
 
         super(Message, self).__init__()
 
+    def is_alive(self):
+        """
+        Indicates whether or not this :class:`nsq.message.Message` instance is timed out based on the
+            msg_timeout parameter.
+        """
+        return not self.timed_out
```
Comment (akellehe) [inline nsq/message.py]: We may want to use `not self.timed_out and not self._has_responded` here for a more comprehensive check

File: nsq/message.py [hunk 9]
```diff
@@ -51,13 +51,22 @@ class Message(event.EventedMixin):
...
+
         self.id = id
         self.body = body
         self.timestamp = timestamp
         self.attempts = attempts
+        self.timed_out = False
 
         super(Message, self).__init__()
 
+    def is_alive(self):
```
Comment (mreiferson) [inline nsq/message.py]: I don't see this method used anywhere?  Do we need it?

File: nsq/message.py [hunk 10]
```diff
@@ -51,13 +51,22 @@ class Message(event.EventedMixin):
     def __init__(self, id, body, timestamp, attempts):
         self._async_enabled = False
         self._has_responded = False
+
         self.id = id
         self.body = body
         self.timestamp = timestamp
         self.attempts = attempts
+        self.timed_out = False
```
Comment (mreiferson) [inline nsq/message.py]: If we don't need `is_alive`, then we don't need this variable either...

File: nsq/async.py [hunk 11]
```diff
@@ -468,9 +471,18 @@ def _on_data(self, data, **kwargs):
             self.in_flight += 1
 
             message = protocol.decode_message(data)
+            if self.msg_timeout:
+                if message.id in self.__timeouts:
```
Comment (mreiferson) [inline nsq/async.py]: Shouldn't it be impossible to have a duplicate message in `self.__timeouts`?  In order to see the same message ID a message would need to have:

1. been explicitly requeued (in which case the timeout callback needs to be cleaned up)
2. timed out and been redelivered (in which case the timeout callback would have fired and been cleaned up)

So, I don't think we need this check?

File: nsq/reader.py [hunk 12]
```diff
@@ -275,6 +282,15 @@ def set_message_handler(self, message_handler):
         assert callable(message_handler), 'message_handler must be callable'
         self.message_handler = message_handler
 
+    def set_message_timeout_handler(self, message_timeout_handler):
```
Comment (mreiferson) [inline nsq/reader.py]: We only need `set_message_handler` because in certain cases you want to pass your `nsq.Reader` instance to your handler, so it needs to be instantiated before you can be passed in.

Is that why we need this `set_message_timeout_handler`?

File: nsq/async.py [hunk 13]
```diff
@@ -468,9 +471,18 @@ def _on_data(self, data, **kwargs):
...
+                    self.io_loop.remove_timeout(self.__timeouts[message.id])
+
+                self.__timeouts[message.id] = self.io_loop.add_timeout(
+                    self.msg_timeout/1000., lambda: self.trigger(event.MESSAGE_TIMEOUT, self, message)
+                )
+
             message.on(event.FINISH, self._on_message_finish)
             message.on(event.REQUEUE, self._on_message_requeue)
             message.on(event.TOUCH, self._on_message_touch)
+            message.on(event.TIMEOUT, self._on_message_timeout)
```
Comment (mreiferson) [inline nsq/async.py]: `s/TIMEOUT/MESSAGE_TIMEOUT`

File: nsq/reader.py [hunk 14]
```diff
@@ -476,6 +499,7 @@ def connect_to_nsqd(self, host, port):
         conn.on('close', self._on_connection_close)
         conn.on('ready', self._on_connection_ready)
         conn.on('message', self._on_message)
+        conn.on('message_timeout', self._on_message_timeout)
```
Comment (mreiferson) [inline nsq/reader.py]: I don't see `AsyncConn` propagating this event, so I don't think this is listening to anything.  Maybe `_on_message_timeout` should be implemented at the `AsyncConn`, not `Reader`, level?

File: nsq/async.py (hunk 11 above)
Comment (akellehe) [inline nsq/async.py]: Yeah, I think you're right

File: nsq/message.py (hunk 9 above)
Comment (akellehe) [inline nsq/message.py]: This method offers a second approach to users for acting on the fact their message has timed out. If, for example, a message is timed out and about to undergo an expensive operation; the user can check `is_alive` and simply discard the message. 

The alternative using callbacks would require some significant overhead: either a registry of messages that have timed out so the same thing can be accomplished, or setting a similar flag in the callback itself. Imho neither of those two are ideal, so offering this interface gives the user a simple alternative.

File: nsq/message.py (hunk 8 above)
Comment (akellehe) [inline nsq/message.py]: Worth noting: I think we could change the name from `is_alive` to `is_in_flight` if we include both flags.

File: nsq/reader.py (hunk 12 above)
Comment (akellehe) [inline nsq/reader.py]: Nope! I was just following the same pattern. I'll change this to a simple `=`

File: nsq/reader.py (hunk 14 above)
Comment (akellehe) [inline nsq/reader.py]: Wouldn't [this trigger][1] propagate to this callback?

[1]: https://github.com/akellehe/pynsq/blob/message-timeout-callback-and-state-accessor/nsq/async.py#L479

Comment (akellehe) [inline nsq/reader.py]: Added [a test][1] to confirm

[1]: https://github.com/akellehe/pynsq/blob/message-timeout-callback-and-state-accessor/tests/test_reader.py#L201-L223

File: nsq/message.py (hunk 9 above)
Comment (mreiferson) [inline nsq/message.py]: OK, but I don't think it needs to be implemented by relying on some parent to set a property on this object - the message could gain the knowledge of its timeout and then just perform some time logic here?

File: nsq/async.py [hunk 15]
```diff
@@ -520,3 +530,20 @@ def _on_message_touch(self, message, **kwargs):
                 conn=self,
                 error=protocol.SendError('failed to send TOUCH %s' % message.id, e),
             )
+
+    def __clear_timeout(self, message):
```
Comment (mreiferson) [inline nsq/async.py]: I think all of these related timeout functions and variables need to be more specifically named, i.e. `s/timeout/message_timeout`

File: nsq/message.py (hunk 9 above)
Comment (akellehe) [inline nsq/message.py]: That's a good point. I'll take a crack at something less leaky.

File: nsq/async.py (hunk 15 above)
Comment (akellehe) [inline nsq/async.py]: +1

**Comment by akellehe** [comment]
@mreiferson I think we're all set. Are these tests adequate? Should we test a few more cases?

**Comment by mreiferson** [comment]
@akellehe Nice work so far!

I have a general concern that for high volume consumers there could be non-trivial overhead maintaining timers for every message received.

As implemented, this is going to track timeouts for all messages even if a message timeout handler _hasn't_ been provided.  Should we only track timeouts when a handler is provided?   If so, that means that `Message.is_in_flight()` needs to track state independent of message timeout events.

**Comment by akellehe** [comment]
@mreiferson these are good points. LMK what you think about this approach to passing the responsibility for handling timeout state to the `nsq.message.Message`

File: nsq/protocol.py [hunk 16]
```diff
@@ -61,12 +61,12 @@ def unpack_response(data):
     return frame, data[4:]
 
 
-def decode_message(data):
+def decode_message(data, timeout=None):
```
Comment (akellehe) [inline nsq/protocol.py]: I don't think setting the timeout here is ideal, `decode_message` isn't very idiomatic when we also use it to pass the message `timeout`. The best alternative I can think of is to keep track of timeouts by default, but we'd see the additional overhead you mentioned.

Comment (mreiferson) [inline nsq/protocol.py]: I think I would prefer a method `_set_timeout` instead?
